contracts/TokenDistributor.sol

```diff
@@ -0,0 +1,212 @@
+
+        uint256 totalStaked = stakingRewardsV2.totalSupplyAtTime(
+            _startOfEpoch(epochNumber)
+        );
+        if (totalStaked == 0) {
+            revert NothingStakedThatEpoch();
+        }
+
+        uint256 proportionalFees = calculateEpochFees(to, epochNumber);
+
+        if (proportionalFees == 0) {
+            revert CannotClaim0Fees();
+        }
+
+        lastTokenBalance -= proportionalFees;
+
+        kwenta.approve(address(rewardEscrowV2), proportionalFees);
+        rewardEscrowV2.createEscrowEntry(to, proportionalFees, 52 weeks, 90);
+
+        emit EpochClaim(to, epochNumber, proportionalFees);
+    }
+
+    /// @inheritdoc ITokenDistributor
+    function claimMany(address to, uint[] memory epochs) public {
+        for (uint i = 0; i < epochs.length; i++) {
+            uint epochNumber = epochs[i];
+            claimEpoch(to, epochNumber);
+        }
+    }
+
+    /// @inheritdoc ITokenDistributor
+    function calculateEpochFees(
+        address to,
+        uint epochNumber
+    ) public view override returns (uint256) {
+        uint epochStart = _startOfEpoch(epochNumber);
+        uint256 userStaked = stakingRewardsV2.balanceAtTime(to, epochStart);
+        uint256 totalStaked = stakingRewardsV2.totalSupplyAtTime(epochStart);
+        if (totalStaked == 0) {
+            return 0;
+        }
+        uint256 proportionalFees = (tokensPerEpoch[epochStart] * userStaked) /
+            totalStaked;
+
+        return proportionalFees;
+    }
+
+    /// @notice function for calculating the start of a week with an offset
+    function _startOfWeek(uint timestamp) internal view returns (uint) {
+        /// @dev remove offset then truncate and then put offset back because
+        /// you cannot truncate to an "offset" time - always truncates to the start
+        /// of unix time -
+        /// @dev this also prevents false truncation: without removing then adding
+        /// offset, the end of a normal week but before the end of an offset week
+        /// will get truncated to the next normal week even though the true week (offset)
+        /// has not ended yet
+        return (((timestamp - offset) / 1 weeks) * 1 weeks) + offset;
+    }
+
+    /// @notice function for calculating the start of an epoch
+    function _startOfEpoch(uint epochNumber) internal view returns (uint) {
+        return (epochNumber * 1 weeks) + startTime;
+    }
+
+    /// @notice function for determining if a checkpoint is necessary
+    function _checkpointWhenReady() internal {
+        /// @dev if more than 24 hours from last checkpoint
+        if ((block.timestamp - lastCheckpoint > 1 days)) {
+            checkpointToken();
+        }
+        /// @dev if it is the first claim of the week then checkpoint.
+        /// this condition is so that the end of a week always
+        /// gets updated before its claimed (even if < 24 hrs)
+        if ((block.timestamp - _startOfWeek(lastCheckpoint)) > 1 weeks) {
```

Comment:
**Issue:** I don't believe there are any tests that check the edge case where `block.timestamp - _startOfWeek(lastCheckpoint) == 1 weeks`, we should test this situation.